File: module/pytimer/pytimer.cc

// Copyright (c) 2018 brinkqiang (brink.qiang@gmail.com)
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.

#include "pytimer.h"
#include <iostream>
#include <atomic>
#include "dmutil.h"

std::atomic_bool CPyTimer::m_bStop{false};

CPyTimer::CPyTimer()
{

}

void CPyTimer::settimer(uint64_t qwIDEvent,  uint64_t qwElapse,
                         pybind11::function f)
{
    SetPyTimer(qwIDEvent, qwElapse, f);
}

void CPyTimer::killtimer(uint64_t qwIDEvent)
{
    KillTimer(qwIDEvent);
}

void CPyTimer::killall()
{
    KillTimer();
}

void CPyTimer::sleepms(uint64_t qwElapse, pybind11::function f)
{
    SetPyTimer(0, qwElapse, f, true);
}

std::string CPyTimer::gettime()
{
    return DMFormatDateTime();
}

void CPyTimer::stop()
{
    m_bStop = true;
}

void CPyTimer::run()
{
    int nEvent = 0;
    bool bBusy = false;

    while (!m_bStop)
    {
        bBusy = false;

        if (CDMTimerModule::Instance()->Run())
        {
            bBusy = true;
        }

        if (!bBusy)
        {
            SleepMs(1);
        }
    }

    CDMTimerModule::Instance()->Run();
}

int CPyTimer::poll()
{
    return CDMTimerModule::Instance()->Run();
}

File: module/pytimer/pytimer.py
# -*- coding: utf-8 -*-
import pytimer

timer = pytimer.CPyTimer()

def main_idle(id):
    print(f"{pytimer.gettime()} hello main_idle: {id}")

timer.settimer(1, 1000, main_idle)

print(pytimer.gettime())

# Lua 风格嵌套回调实现
timer.sleepms(2000, lambda : (
    (lambda count=[0]: (  # 闭包捕获可变对象
        print(f"sleepms 2000 {pytimer.gettime()}"),
        timer.settimer(2, 1000, lambda id: (
            (lambda count=count: (
                count.__setitem__(0, count[0] + 1),
                print(f"hello world {count[0]}"),
                count[0] >= 10 and (
                    timer.killtimer(id),
                    timer.killall(),
                    pytimer.stop()
                )
            ))()
        ))
    ))()
))

pytimer.run()